src/advanced/JupyterNotebook/index.tsx

```diff
@@ -0,0 +1,322 @@
+                    {bodyColumnContent}
+                </Paper>
+            </Grid>
+        </>
+    );
+};
+
+// -- renders a cell of type 'markdown'
+const JupyterNotebookMarkdownCell: React.FC<JupyterNotebookCellProps> = ({
+    cell,
+    executionCount: execution_count,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={''}
+            bodyColumnContent={<ReactMarkdown children={sourceLines(cell.source)} />}
+            bodyColumnStyle={{ padding: '0px 8px 0px 8px', margin: '0px 0px', border: '1px solid #AAA' }}
+        />
+    );
+};
+
+// -- renders a cell of type 'heading'
+const JupyterNotebookHeadingCell: React.FC<JupyterNotebookCellProps> = ({
+    cell,
+    executionCount: execution_count,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={''}
+            bodyColumnContent={
+                <Container
+                    style={{
+                        marginBottom: '0px',
+                        boxShadow: 'none',
+                    }}
+                >
+                    <Heading variant="h1">{cell.source[0]}</Heading>
+                </Container>
+            }
+            bodyColumnStyle={{ padding: '0px 8px 0px 8px', margin: '0px 0px', border: '1px solid #AAA' }}
+        />
+    );
+};
+
+// -- renders a cell of type 'stream'
+const JupyterNotebookOutputStreamCell: React.FC<JupyterNotebookOutputCellProps> = ({
+    cell,
+    executionCount,
+    output,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={`Out [${executionCount}]`}
+            bodyColumnContent={
+                <ReactMarkdown children={'**[' + output.name + ']**\n```\n ' + sourceLines(output.text) + '```'} />
+            }
+            bodyColumnStyle={{
+                padding: '0px 8px 0px 8px',
+                margin: '0px 0px',
+                backgroundColor: output.name === 'stderr' ? '#FDD' : '#DFD',
+                border: '1px solid #AAA',
+            }}
+        />
+    );
+};
+
+// -- renders a cell of type 'error'
+const JupyterNotebookOutputErrorCell: React.FC<JupyterNotebookOutputCellProps> = ({
+    cell,
+    executionCount,
+    output,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={`In [${executionCount}]`}
+            bodyColumnContent={
+                <ReactMarkdown
+                    children={
+                        '**[' +
+                        output.evalue +
+                        ']**\n```\n ' +
+                        sourceLines(output.traceback).replace(
+                            /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, // strip out ansi control sequences common in output traces
+                            ''
+                        ) +
+                        '\n```'
+                    }
```

Comment:
it will be better to move this to constant out from `render`

Reply:
yes it would